lib/async_hooks.js

```diff
@@ -252,10 +252,65 @@ class AsyncResource {
   }
 }
 
+let _tls_wrap;
+
+function lazyLoadTlsWrap() {
+  _tls_wrap ??= require('_tls_wrap');
+}
+
+let dgram;
+
+function lazyLoadDgram() {
+  dgram ??= require('dgram');
+}
+
+let internal_dgram;
+
+function lazyLoadInternalDgram() {
+  internal_dgram ??= require('internal/dgram');
+}
+
+let net;
+
+function lazyLoadNet() {
+  net ??= require('net');
+}
+
+function lookupResourceWithStorage() {
+  // When a TCP/UDP object is created, its owner_symbol is uninitialized, so
+  // calling executionAsyncResource() would return the TCP/UDP object itself.
+  // So the store is initially saved on the TCP/UDP object. Later on, when the
+  // Socket/TLSSocket object is created, the TCP/UDP object is wrapped by the
+  // Socket/TLSSocket object in the _handle property and the owner_symbol of the
+  // TCP/UDP object is set to the corresponding Socket/TLSSocket object. So
+  // calling executionAsyncResource() now would return the Socket/TLSSocket
+  // object, which does not contain the storage that was saved initially. Hence,
+  // in the case of a Socket/TLSSocket object, we must use the underlying
+  // TCP/UDP object, if available, for storage.
+
+  lazyLoadDgram();
+  lazyLoadInternalDgram();
+  lazyLoadNet();
+  lazyLoadTlsWrap();
+
+  let resource = executionAsyncResource();
+
+  if ((resource instanceof net.Socket ||
+       resource instanceof _tls_wrap.TLSSocket) &&
+      resource._handle != null) {
+    resource = resource._handle;
+  } else if (resource instanceof dgram.Socket &&
+             resource[internal_dgram.kStateSymbol].handle != null) {
+    resource = resource[internal_dgram.kStateSymbol].handle;
+  }
+
+  return resource;
+}
+
 const storageList = [];
 const storageHook = createHook({
   init(asyncId, type, triggerAsyncId, resource) {
-    const currentResource = executionAsyncResource();
+    const currentResource = lookupResourceWithStorage();
```

Comment:
Does this mean the API [`executionAsyncResource()`](https://nodejs.org/dist/latest-v16.x/docs/api/async_hooks.html#async_hooksexecutionasyncresource) are our our of sync now?

Reply:
@Flarna in the cases where executionAsyncResource() returns a net/dgram.Socket or a TLSSocket, the AsyncLocalStorage is saved inside the handle, i.e., the TCP/UDP object. Otherwise, it is still in sync.

Reply:
I think it should be always in sync. I don't think that `AsyncLocalStore` should have a different opinion on what is the current resource than other parts of node using AsyncHooks directly (e.g. domains).

Reply:
@Flarna I've moved lookupResourceWithStorage() into executionAsyncResource() in this commit: https://github.com/nodejs/node/pull/40741/commits/9d8b2348af329bd31d40ddafc44c5cb5ca179e63, PTAL